Mathlib/Util/MemoFix.lean

```diff
@@ -0,0 +1,32 @@
+/-
+Copyright (c) 2022 Gabriel Ebner. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Gabriel Ebner, E.W.Ayers
+-/
+import Std
+
+open Std ShareCommon
+
+/-!
+# Fixpoint function with memoisation
+
+-/
+
+private unsafe def memoFixImpl {α β : Type} [Inhabited β] (f : (α → β) → (α → β)) (a : α) : β := unsafeBaseIO do
```

Comment:
One more thing:
```suggestion
private unsafe def memoFixImpl {α : Type u} {β : Type v} [Inhabited β] (f : (α → β) → (α → β)) (a : α) : β := unsafeBaseIO do
```

Reply:
Unfortunately IO.Ref is not universe polymorphic so this doesn't work. I can make the constant univ-poly though?

Reply:
You could try `ObjectMap` and `unsafeCast` the values as well.

Reply:
Sorry I'm struggling to get this to work

Reply:
Oh, that seems to require a bit more restructuring, since unsafeBaseIO is limited to Type 0 as well... Maybe the easiest way out is just to do whole implementation first for Object, and then to cast the whole function afterwards?
```lean
private unsafe def memoFixImplObj (f : (Object → Object) → (Object → Object)) (a : Object) : Object := unsafeBaseIO do
  let cache : IO.Ref ObjectMap ← ST.mkRef (mkObjectMap ())
  let rec fix (a) := unsafeBaseIO do
    if let some b := (← cache.get).find? a then
      return b
    let b := f fix a
    cache.modify (·.insert a b)
    pure b
  pure $ fix a

private unsafe def memoFixImpl {α : Type u} {β : Type v} [Nonempty β] : (f : (α → β) → (α → β)) → (a : α) → β :=
  unsafeCast memoFixImplObj
```

Reply:
Nice thanks. I tried doing something like this but I messed something up and it would hang.